### ISSUE TITLE
not connecting to dartls

### ISSUE THREAD

**Issue by pxsanghyo** [akinsho/flutter-tools.nvim#73]
I'm having problem with not connecting to lsp.

```lua
if O.lang.dart.flutter_tools.active then
  require('flutter-tools').setup{}
else
  require("lspconfig").dartls.setup {
    cmd = { "dart", O.lang.dart.sdk_path, "--lsp" },
    on_attach = require("lsp").common_on_attach,
    init_options = {
      closingLabels = false,
      flutterOutline = false,
      onlyAnalyzeProjectsWithOpenFiles = false,
      outline = false,
      suggestFromUnimportedLibraries = true,
    },
  }
end
```

```lua
  use {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    disable = not O.lang.dart.flutter_tools.active,
  }
```

I have flutter set in $PATH so I just left it default.
It worked once before but for some reason it stopped working.

**Comment by akinsho** [comment]
@pxsanghyo can you check what your lsp logs say. If you don't know where those are you can find the path by checking `:lua vim.lsp.log.get_filename()` and see if it indicates whether or not there are any errors.

From this issue I can't tell what OS system you use, what version/channel of dart you are on or any other information to help.

**Comment by pxsanghyo** [comment]
@akinsho I'm not sure if this is correct log but I found it in ~/.cache/nvim/lsp.log 
```
[ START ] 2021-07-05T20:06:01+0900 ] LSP logging initiated
[ START ] 2021-07-05T20:05:47+0900 ] LSP logging initiated
```
Here is my flutter doctor output
```
Flutter (Channel beta, 2.2.0-10.3.pre, on macOS 11.4 20F71 darwin-x64, locale en
```

**Comment by akinsho** [comment]
@pxsanghyo where are you calling the setup block for the plugin? you are using packer it looks like but not using the `config` block to call setup for the plugin.
Can you try something simpler in your setup like

```lua
use {
  "akinsho/flutter-tools.nvim",
  requires = "nvim-lua/plenary.nvim"
  config = function()
   require('flutter-tools').setup{}
  end,
}
```
also can you paste the result of `which flutter`, as well.

**Comment by pxsanghyo** [comment]
```
which flutter
/Users/sanghyo/.flutter/bin/flutter
```

and yep I'm using packer, I tried the config before but didn't work...
let me try again and get back to you

**Comment by pxsanghyo** [comment]
Tried again with `config` block but no go...
Also, no errors in lsp.log

**Comment by akinsho** [comment]
@pxsanghyo did you run `PackerCompile` to ensure everything was now being loaded, you can verify this by adding a print statement in your config block just to make sure it is running. Also just to clarify are you entering a dart file and waiting at least 10seconds because it can be quite slow to startup?

**Comment by pxsanghyo** [comment]
I do `PackerCompile` and then restart `nvim` and open a dart file.
Just now tried waiting for a bit longer to see if it was just slow to startup but it's been few minutes now...

`PackerStatus` shows `flutter-tools` is loaded... sorry I wish I could be more help to find the problem...
It just seems like it should work but it doesn't connect.

**Comment by akinsho** [comment]
@pxsanghyo does `lspconfig` on its own work. Just to verify that your flutter installation can correctly launch the lsp.

**Comment by pxsanghyo** [comment]
I'm using a forked repo LunarVim and creating a pr for `flutter-tools` support.
`https://github.com/pxsanghyo/LunarVim/tree/flutter`

Here are the files I edited. I removed some for brevity.

`/ftplugin/dart.lua`
```
if O.lang.dart.flutter_tools.active then
  require("flutter-tools").setup{}
else
  require("lspconfig").dartls.setup {
    cmd = { "dart", O.lang.dart.sdk_path, "--lsp" },
    on_attach = require("lsp").common_on_attach,
    init_options = {
      closingLabels = false,
      flutterOutline = false,
      onlyAnalyzeProjectsWithOpenFiles = false,
      outline = false,
      suggestFromUnimportedLibraries = true,
    },
  }
end

if O.lang.dart.autoformat then
  require('lv-utils').define_augroups({
    _dart_autoformat = {
      {'BufWritePre', '*.dart', 'lua vim.lsp.buf.formatting_sync(nil, 1000)'}
    }
  })
end
```

`/lua/default-config.lua`
```
CONFIG_PATH = vim.fn.stdpath "config"
DATA_PATH = vim.fn.stdpath "data"
CACHE_PATH = vim.fn.stdpath "cache"
TERMINAL = vim.fn.expand "$TERMINAL"

O = {
  lang = {
    dart = {
      sdk_path = '/usr/lib/dart/bin/snapshots/analysis_server.dart.snapshot',
      autoformat = false,
      flutter_tools = {
        active = false
      }
    }
  }
}
```

`/lua/plugins.lua`
```
local execute = vim.api.nvim_command
local fn = vim.fn

local install_path = fn.stdpath "data" .. "/site/pack/packer/start/packer.nvim"

if fn.empty(fn.glob(install_path)) > 0 then
  execute("!git clone https://github.com/wbthomason/packer.nvim " .. install_path)
  execute "packadd packer.nvim"
end

local packer_ok, packer = pcall(require, "packer")
if not packer_ok then
  return
end

packer.init {
  -- compile_path = vim.fn.stdpath('data')..'/site/pack/loader/start/packer.nvim/plugin/packer_compiled.vim',
  compile_path = require("packer.util").join_paths(vim.fn.stdpath "config", "plugin", "packer_compiled.vim"),
  git = { clone_timeout = 300 },
  display = {
    open_fn = function()
      return require("packer.util").float { border = "single" }
    end,
  },
}

vim.cmd "autocmd BufWritePost plugins.lua PackerCompile"

return require("packer").startup(function(use)
  -- Flutter tools
  use {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    disable = not O.lang.dart.flutter_tools.active,
  }
end)
```

**Comment by pxsanghyo** [comment]
> @pxsanghyo does `lspconfig` on its own work. Just to verify that your flutter installation can correctly launch the lsp.

Yes, dartls works when `lang.dart.flutter_tools.active` is false and when I to set `lang.dart.sdk_path` to my own path which is something like `~/.flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot`

**Comment by akinsho** [comment]
@pxsanghyo tbh the config is a little intricate. A few things to note.
* Loading in a ftplugin is actually not a good idea. This plugin already tries to be minimal by creating an autocommand on setup to *eventually* load when `*.dart` is entered. By loading in a ftplugin this mean the plugin will not have setup it's autocommands when you enter a dart file.This is probably why it is not working.
* Ftplugin files are run *everytime* a dart file is opened and the ft is set, so you will be calling setup on this plugin potentially dozens of times.

Can you please try to reproduce this issue with a minimal init.vim, just one file with the setup I pasted above

**Comment by pxsanghyo** [comment]
Ok I'll try that and get back to you. Thanks

**Comment by pxsanghyo** [comment]
@akinsho I got it working by removing setup call in dart ftplugin.
Lunarvim uses ftplugin heavily so theres no going around that.
atleast flutter-tools is working now.
Thanks for the help.

**Comment by akinsho** [comment]
@pxsanghyo, glad you found a fix, so where did you end up calling setup in the end?

**Comment by pxsanghyo** [comment]
This is what's working for me now.

plugins.lua
```lua
  use {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    config = function() require("flutter-tools").setup{} end,
    disable = not O.lang.dart.flutter_tools.active,
  }
```

dart.lua
```lua
if not O.lang.dart.flutter_tools.active then
  require("lspconfig").dartls.setup {
    cmd = { "dart", O.lang.dart.sdk_path, "--lsp" },
    on_attach = require("lsp").common_on_attach,
    init_options = {
      closingLabels = false,
      flutterOutline = false,
      onlyAnalyzeProjectsWithOpenFiles = false,
      outline = false,
      suggestFromUnimportedLibraries = true,
    },
  }
end

if O.lang.dart.autoformat then
  require('lv-utils').define_augroups({
    _dart_autoformat = {
      {'BufWritePre', '*.dart', 'lua vim.lsp.buf.formatting_sync(nil, 1000)'}
    }
  })
end
```

added packer config to setup AND remove setup from ftplugin.

**Comment by akinsho** [comment]
Thanks maybe this will be helpful to see for any other lunar users 👍🏾

**Comment by barun108** [comment]
> This is what's working for me now.
> 
> plugins.lua
> 
> ```lua
>   use {
>     "akinsho/flutter-tools.nvim",
>     requires = "nvim-lua/plenary.nvim",
>     config = function() require("flutter-tools").setup{} end,
>     disable = not O.lang.dart.flutter_tools.active,
>   }
> ```
> 
> dart.lua
> 
> ```lua
> if not O.lang.dart.flutter_tools.active then
>   require("lspconfig").dartls.setup {
>     cmd = { "dart", O.lang.dart.sdk_path, "--lsp" },
>     on_attach = require("lsp").common_on_attach,
>     init_options = {
>       closingLabels = false,
>       flutterOutline = false,
>       onlyAnalyzeProjectsWithOpenFiles = false,
>       outline = false,
>       suggestFromUnimportedLibraries = true,
>     },
>   }
> end
> 
> if O.lang.dart.autoformat then
>   require('lv-utils').define_augroups({
>     _dart_autoformat = {
>       {'BufWritePre', '*.dart', 'lua vim.lsp.buf.formatting_sync(nil, 1000)'}
>     }
>   })
> end
> ```
> 
> added packer config to setup AND remove setup from ftplugin.

Can you share your all file's configurations for Flutter tools with LunarVim ?

I am still confused

**Comment by pxsanghyo** [comment]
ftplugin/dart.lua
- Replace everything with the following code for `dart.lua`
```lua
if not O.lang.dart.flutter_tools.active then
  require("lspconfig").dartls.setup {
    cmd = { "dart", O.lang.dart.sdk_path, "--lsp" },
    on_attach = require("lsp").common_on_attach,
    init_options = {
      closingLabels = false,
      flutterOutline = false,
      onlyAnalyzeProjectsWithOpenFiles = false,
      outline = false,
      suggestFromUnimportedLibraries = true,
    },
  }
end

if O.lang.dart.autoformat then
  require('lv-utils').define_augroups({
    _dart_autoformat = {
      {'BufWritePre', '*.dart', 'lua vim.lsp.buf.formatting_sync(nil, 1000)'}
    }
  })
end
```

lua/default-config.lua
- Look for `dart = {...}` and add the missing properties: autoformat, flutter_tools
```lua
    dart = {
      sdk_path = '/usr/lib/dart/bin/snapshots/analysis_server.dart.snapshot',
      autoformat = false,
      flutter_tools = {
        active = false
      }
    },
```

lua/plugins.lua
- Somewhere in the `return require("packer").startup(function(use)` block, where plugins are defined, add the following code
```lua
  -- Flutter tools
  use {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    config = function() require("flutter-tools").setup{} end,
    disable = not O.lang.dart.flutter_tools.active,
  }
```

lv-config.dart
- Add the following to turn on flutter_tools (and autoformat if you want)
- sdk_path is not required and is ignored when flutter_tools is active
```lua
-- dart
O.lang.dart.sdk_path = '/Users/pxsanghyo/.flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot'
O.lang.dart.flutter_tools.active = true
O.lang.dart.autoformat = true
```

Let me know if you need any more help

**Comment by derrikfleming** [comment]
@pxsanghyo is there anyway you could post your LV config... I've been trying to get flutter-tools working for weeks to no avail. It works with the first buffer the first time it is opened but not with any others. I've tried with more minimal setups and can't get anywhere with it.

**Comment by spareblock** [comment]
@derrikfleming I'm not doing anything much.

```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "dartls" })

lvim.plugins = {
  {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    config = function()
      require("flutter-tools").setup {
        lsp = {
          on_attach = require("lvim.lsp").common_on_attach
        }
      }
    end,
  },
}
```

Hopefully that works. Good luck

**Comment by pxsanghyo** [comment]
@derrikfleming I'm not doing anything much.

```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "dartls" })

lvim.plugins = {
  {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    config = function()
      require("flutter-tools").setup {
        lsp = {
          on_attach = require("lvim.lsp").common_on_attach
        }
      }
    end,
  },
}
```

Hopefully that works. Good luck
